src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -146,8 +160,31 @@ private Cci.ITypeReference GetStorageStruct(uint size)
             return new ExplicitSizeStruct(size, this, _systemValueType);
         }
 
+        internal bool HasStaticConstructor => _staticConstructor != null;
+
+        internal Cci.IFieldReference GetOrAddModuleVersionId(Cci.ITypeReference mvidType, Func<Cci.IMethodDefinition> staticConstructorCreator)
+        {
+            if (_mvidField == null)
+            {
+                Interlocked.CompareExchange(ref _mvidField, new ModuleVersionIdField("MVID", this, mvidType), null);
+                EnsureStaticConstructorExists(staticConstructorCreator);
+            }
+
+            return _mvidField;
+        }
 
-        // Add a new synthesized method indexed by it's name if the method isn't already present.
+        private void EnsureStaticConstructorExists(Func<Cci.IMethodDefinition> creator)
```

Comment:
As an alternative, we can let the derived type to track the fact whether the static constructor has been created, then it can simply call TryAddSynthesizedMethod to add to the list and we wouldn't have to deal with delegates.